File: src/services/ingestion/extractors/image.test.ts
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, it } from "node:test";
import assert from "node:assert/strict";

import { ContentType, DocumentStatus } from "../../../types/enums.js";
import type { Document } from "../../../types/models.js";
import { ImageExtractor } from "./image.js";

function makeDocument(overrides: Partial<Document> = {}): Document {
  return {
    id: "doc-1",
    title: "Image Doc",
    contentType: ContentType.Image,
    rawContent: "",
    containerTag: "default",
    metadata: {},
    status: DocumentStatus.Queued,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
    ...overrides
  };
}

describe("ImageExtractor", () => {
  it("extract trims OCR text", async () => {
    const extractor = new ImageExtractor() as ImageExtractor & {
      getImageBuffer: (doc: Document) => Promise<Buffer>;
      loadTesseract: () => Promise<{ recognize: (image: Buffer, language?: string) => Promise<{ data?: { text?: string } }> }>;
    };

    extractor.getImageBuffer = async () => Buffer.from("mock-image", "utf8");
    extractor.loadTesseract = async () => ({
      recognize: async (image: Buffer, language?: string) => {
        assert.equal(image.toString("utf8"), "mock-image");
        assert.equal(language, "eng");
        return { data: { text: "  detected text  \n" } };
      }
    });

    const result = await extractor.extract(makeDocument());
    assert.equal(result, "detected text");
  });

  it("reads image bytes from filePath", async () => {
    const tempDir = await mkdtemp(join(tmpdir(), "image-extractor-test-"));
    const filePath = join(tempDir, "sample.png");

    try {
      const payload = Buffer.from([0x89, 0x50, 0x4e, 0x47]);
      await writeFile(filePath, payload);

      const extractor = new ImageExtractor() as ImageExtractor & {
        getImageBuffer: (doc: Document) => Promise<Buffer>;
      };

      const buffer = await extractor.getImageBuffer(makeDocument({ filePath }));
      assert.equal(buffer.equals(payload), true);
    } finally {
      await rm(tempDir, { recursive: true, force: true });
    }
  });

  it("accepts image data URL in rawContent", async () => {
    const extractor = new ImageExtractor() as ImageExtractor & {
      getImageBuffer: (doc: Document) => Promise<Buffer>;
    };

    const bytes = Buffer.from([0x89, 0x50, 0x4e, 0x47]);
    const rawContent = `data:image/png;base64,${bytes.toString("base64")}`;

    const buffer = await extractor.getImageBuffer(makeDocument({ rawContent }));
    assert.equal(buffer.equals(bytes), true);
  });

  it("accepts base64 encoded rawContent", async () => {
    const extractor = new ImageExtractor() as ImageExtractor & {
      getImageBuffer: (doc: Document) => Promise<Buffer>;
    };

    const bytes = Buffer.from("image-bytes", "utf8");
    const buffer = await extractor.getImageBuffer(
      makeDocument({ rawContent: bytes.toString("base64") })
    );

    assert.equal(buffer.toString("utf8"), "image-bytes");
  });

  it("falls back to utf8 rawContent when not base64", async () => {
    const extractor = new ImageExtractor() as ImageExtractor & {
      getImageBuffer: (doc: Document) => Promise<Buffer>;
    };

    const rawContent = "plain image description";
    const buffer = await extractor.getImageBuffer(makeDocument({ rawContent }));

    assert.equal(buffer.toString("utf8"), rawContent);
  });

  it("throws when rawContent is missing and filePath is not provided", async () => {
    const extractor = new ImageExtractor();

    await assert.rejects(
      () => extractor.extract(makeDocument({ rawContent: "" })),
      /Image extractor requires document\.filePath or rawContent/
    );
  });

  it("throws when rawContent is empty after trimming", async () => {
    const extractor = new ImageExtractor();

    await assert.rejects(
      () => extractor.extract(makeDocument({ rawContent: "   \n\t   " })),
      /Image rawContent was empty/
    );
  });

  it("wraps file read errors with file path context", async () => {
    const extractor = new ImageExtractor();
    const missingPath = "/tmp/definitely-missing-image-file.png";

    await assert.rejects(
      () => extractor.extract(makeDocument({ filePath: missingPath })),
      new RegExp(`Failed to read image file at ${missingPath}`)
    );
  });
});

File: src/services/ingestion/extractors/image_test.py
import base64
import pytest
from pathlib import Path
from unittest.mock import AsyncMock, patch
from src.services.ingestion.extractors.image import ImageExtractor
from src.types.models import Document
from src.types.enums import ContentType, DocumentStatus


def make_document(overrides: dict | None = None) -> Document:
    """Helper to create a test document with defaults."""
    defaults = {
        "id": "doc-1",
        "title": "Image Doc",
        "content_type": ContentType.Image,
        "raw_content": "",
        "container_tag": "default",
        "metadata": {},
        "status": DocumentStatus.Queued,
        "file_path": None,
    }
    if overrides:
        defaults.update(overrides)
    return Document(**defaults)


@pytest.fixture
def extractor() -> ImageExtractor:
    return ImageExtractor()


@pytest.fixture
def sample_image_bytes() -> bytes:
    """PNG file signature followed by minimal content."""
    return bytes([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]) + b"mock-image-data"


@pytest.mark.asyncio
async def test_extract_trims_ocr_text(extractor: ImageExtractor) -> None:
    """Test that extracted OCR text is properly trimmed."""
    mock_tesseract = AsyncMock()
    mock_tesseract.recognize.return_value = {"data": {"text": "  detected text  \n"}}

    with patch.object(extractor, "_load_tesseract", return_value=mock_tesseract):
        with patch.object(extractor, "_get_image_buffer", return_value=b"mock-image"):
            doc = make_document({"raw_content": "base64encoded"})
            result = await extractor.extract(doc)

    assert result == "detected text"
    mock_tesseract.recognize.assert_called_once()
    call_args = mock_tesseract.recognize.call_args
    assert call_args[0][0] == b"mock-image"
    assert call_args[1].get("language") == "eng"


@pytest.mark.asyncio
async def test_reads_image_bytes_from_file_path(
    extractor: ImageExtractor, tmp_path: Path, sample_image_bytes: bytes
) -> None:
    """Test reading image bytes from a file path."""
    image_file = tmp_path / "sample.png"
    image_file.write_bytes(sample_image_bytes)

    doc = make_document({"file_path": str(image_file)})
    buffer = await extractor._get_image_buffer(doc)

    assert buffer == sample_image_bytes


@pytest.mark.asyncio
async def test_accepts_image_data_url_in_raw_content(
    extractor: ImageExtractor, sample_image_bytes: bytes
) -> None:
    """Test handling of data:image/png;base64 URLs."""
    b64_content = base64.b64encode(sample_image_bytes).decode("utf-8")
    raw_content = f"data:image/png;base64,{b64_content}"

    doc = make_document({"raw_content": raw_content})
    buffer = await extractor._get_image_buffer(doc)

    assert buffer == sample_image_bytes


@pytest.mark.asyncio
async def test_accepts_base64_encoded_raw_content(
    extractor: ImageExtractor, sample_image_bytes: bytes
) -> None:
    """Test handling of plain base64 encoded content."""
    b64_content = base64.b64encode(sample_image_bytes).decode("utf-8")

    doc = make_document({"raw_content": b64_content})
    buffer = await extractor._get_image_buffer(doc)

    assert buffer == sample_image_bytes


@pytest.mark.asyncio
async def test_falls_back_to_utf8_raw_content_when_not_base64(
    extractor: ImageExtractor,
) -> None:
    """Test that non-base64 content is treated as UTF-8 text."""
    raw_content = "plain image description"

    doc = make_document({"raw_content": raw_content})
    buffer = await extractor._get_image_buffer(doc)

    assert buffer.decode("utf-8") == raw_content


@pytest.mark.asyncio
async def test_throws_when_raw_content_missing_and_file_path_not_provided(
    extractor: ImageExtractor,
) -> None:
    """Test error when neither raw_content nor file_path is provided."""
    doc = make_document({"raw_content": "", "file_path": None})

    with pytest.raises(
        ValueError, match="Image extractor requires document.file_path or rawContent"
    ):
        await extractor.extract(doc)


@pytest.mark.asyncio
async def test_throws_when_raw_content_is_empty_after_trimming(
    extractor: ImageExtractor,
) -> None:
    """Test error when raw_content contains only whitespace."""
    doc = make_document({"raw_content": " \n\t "})

    with pytest.raises(ValueError, match="Image rawContent was empty"):
        await extractor.extract(doc)


@pytest.mark.asyncio
async def test_wraps_file_read_errors_with_file_path_context(
    extractor: ImageExtractor,
) -> None:
    """Test that file read errors include the file path in the message."""
    missing_path = "/tmp/definitely-missing-image-file.png"
    doc = make_document({"file_path": missing_path})

    with pytest.raises(
        RuntimeError, match=f"Failed to read image file at {missing_path}"
    ):
        await extractor.extract(doc)
